services/src/main/java/org/keycloak/authentication/authenticators/browser/WebAuthnPasswordlessAuthenticator.java

```diff
@@ -0,0 +1,75 @@
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.keycloak.authentication.authenticators.browser;
+
+import java.util.Collections;
+import java.util.List;
+
+import org.keycloak.authentication.AuthenticationFlowContext;
+import org.keycloak.authentication.RequiredActionFactory;
+import org.keycloak.authentication.RequiredActionProvider;
+import org.keycloak.authentication.requiredactions.WebAuthnPasswordlessRegisterFactory;
+import org.keycloak.credential.CredentialProvider;
+import org.keycloak.credential.WebAuthnPasswordlessCredentialProvider;
+import org.keycloak.credential.WebAuthnPasswordlessCredentialProviderFactory;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.models.WebAuthnPolicy;
+import org.keycloak.models.credential.WebAuthnCredentialModel;
+
+/**
+ * Authenticator for WebAuthn authentication with passwordless credential. This class is temporary and will be likely
+ * removed in the future during future improvements in authentication SPI
+ */
+public class WebAuthnPasswordlessAuthenticator extends WebAuthnAuthenticator {
+
+    public WebAuthnPasswordlessAuthenticator(KeycloakSession session) {
+        super(session);
+    }
+
+    @Override
+    protected WebAuthnPolicy getWebAuthnPolicy(AuthenticationFlowContext context) {
+        return context.getRealm().getWebAuthnPolicyPasswordless();
+    }
+
+    @Override
+    protected String getCredentialType() {
+        return WebAuthnCredentialModel.TYPE_PASSWORDLESS;
+    }
+
+    @Override
+    public void setRequiredActions(KeycloakSession session, RealmModel realm, UserModel user) {
+        // ask the user to do required action to register webauthn authenticator
+        if (!user.getRequiredActions().contains(WebAuthnPasswordlessRegisterFactory.PROVIDER_ID)) {
+            user.addRequiredAction(WebAuthnPasswordlessRegisterFactory.PROVIDER_ID);
+        }
+    }
+
+    @Override
+    public List<RequiredActionFactory> getRequiredActions(KeycloakSession session) {
+        return Collections.singletonList((WebAuthnPasswordlessRegisterFactory)session.getKeycloakSessionFactory().getProviderFactory(RequiredActionProvider.class, WebAuthnPasswordlessRegisterFactory.PROVIDER_ID));
+    }
+
+
+    public WebAuthnPasswordlessCredentialProvider getCredentialProvider(KeycloakSession session) {
```

Comment:
It seems that no one calls this method `getCredentialProvider`. IMO, it might be better to delete this method.

Reply:
Thanks, the only reason why I kept this method was, that I originally worked on this on top of the other branch for introduce "Try another way" link. That task changed some stuff in WebAuthnAuthenticator class (added CredentialValidator as an implementing interface among other things).

Now that other PR is merged in Keycloak master, which means that the method became to be used :) I needed to do few more very small changes in the WebAuthnAuthenticator and WebAuthnPasswordlessAuthenticator class. Do you have a chance to re-review?

Reply:
Yes, I'll try to re-review.